File: algorithms/js/removeElement/removeElement.js
// Source : https://leetcode-cn.com/problems/remove-element/
// Author : Tian Zhuo
// Date   : 2018-11-24

/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
    let i = 0;
    let j = nums.length - 1;
    while(i <= j){
        if(nums[i] == val && nums[j] != val){
            nums[i] = nums[j];
            nums[j] = null;
            j--;
            i++;
        }else if(nums[i] == val && nums[j] == val){
            nums[j] = null;
            j--;
        }else if(nums[i] != val){
            i++;
        }
    }
    return j + 1;
};
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
let removeElement1 = function(nums, val) {
    let i = 0;
    let j = nums.length - 1;
    while (i <= j) {
        if(nums[i] === val){
            [nums[i], nums[j]] = [nums[j], nums[i]]
            --j;
        } else {
            ++i;
        }
    }
    return i;
};

File: src/main/java/removeElement/RemoveElement.java
/*
 * @Source: https://leetcode-cn.com/problems/remove-element/
 * @Author: Tian Zhuo
 * @Date: 2018-12-22 15:16:57
 */
package removeElement;

public class RemoveElement {
    public int removeElement(int[] nums, int val) {
        int i = 0;
        for (int j = 0; j < nums.length; j++) {
            if(i != j && nums[j] != val){
                nums[i] = nums[j];
                nums[j] = val;
                i++;
            }
            if(nums[i] != val){
                i++;
            }
        }
        return i;
    }
}